### ISSUE TITLE
DB::Exception: Unsupported type Decimal(13, 4) when using MaterializeMySQL

### ISSUE THREAD

**Issue by liangli1213** [ClickHouse/ClickHouse#14157]
It seems MaterializeMySQL doesn't support Decimal data type when doing replication.

**Comment by liangli1213** [comment]
@BohuTANG @zhang2014 , need your confirmation.

**Comment by l1t1** [comment]
i got the same problem
first create a table in mysql
```
mysql> create table dec2(a int, b decimal(10,2), primary key(a));
Query OK, 0 rows affected (0.19 sec)

mysql> insert into dec2 values (1,1.2),(2,2.4);
Query OK, 2 rows affected (0.01 sec)
Records: 2  Duplicates: 0  Warnings: 0

mysql> select * from dec2;
+---+------+
| a | b    |
+---+------+
| 1 | 1.20 |
| 2 | 2.40 |
+---+------+

```
then in clickhouse
```
SET allow_experimental_database_materialize_mysql=1;
CREATE DATABASE ckdb2 ENGINE = MaterializeMySQL('127.0.0.1:3306', 'ckdb2', 'root', 'A123b_456');
SHOW TABLES;

┌─name─┐
│ dec2 │
└──────┘
 select * from dec2;

Received exception from server (version 20.8.1):
Code: 50. DB::Exception: Received from localhost:9000. DB::Exception: Unsupported type Decimal(10, 2).

```

**Comment by l1t1** [comment]
the table is created
```
DESCRIBE TABLE dec2

┌─name─────┬─type─────────────────────┬─default_type─┬─default_expression─┬─comment─┬─codec_expression─┬─ttl_expression─┐
│ a        │ Int32                    │              │                    │         │                  │                │
│ b        │ Nullable(Decimal(10, 2)) │              │                    │         │                  │                │
│ _sign    │ Int8                     │ MATERIALIZED │ 1                  │         │                  │                │
│ _version │ UInt64                   │ MATERIALIZED │ 1                  │         │                  │                │
└──────────┴──────────────────────────┴──────────────┴────────────────────┴─────────┴──────────────────┴────────────────┘
```

**Comment by liangli1213** [comment]
@l1t1 , table has been created, but data sync failed.

**Comment by l1t1** [comment]
i found two types of tables arenot supported, are there any other more ?
https://github.com/ClickHouse/ClickHouse/issues/14141

**Comment by dengc367** [comment]
me too. decimal type is not supported when using the MaterializeMySQL engine. @filimonov @zhang2014 @BohuTANG 

the stacktrace is below:

2020.08.28 00:01:42.718954 [ 7739 ] {cf4473a4-8b5a-4c3f-8aab-5f642e9bbf03} <Debug> executeQuery: (from [::ffff:127.0.0.1]:33014) select * from price_print;
2020.08.28 00:01:42.719716 [ 7739 ] {cf4473a4-8b5a-4c3f-8aab-5f642e9bbf03} <Error> executeQuery: Code: 50, e.displayText() = DB::Exception: Unsupported type Decimal(12, 2) (version 20.8.1.4447 (official build)) (from [::ffff:127.0.0.1]:33014) (in query: select * from price_print;), Stack trace (when copying this message, always include the lines below):

0. Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x1a708040 in /usr/bin/clickhouse
1. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0xff0e78d in /usr/bin/clickhouse
2. ? @ 0x16e33687 in /usr/bin/clickhouse
3. DB::MySQLBlockInputStream::MySQLBlockInputStream(mysqlxx::Pool::Entry const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::Block const&, unsigned long, bool) @ 0x17dab3d8 in /usr/bin/clickhouse
4. ? @ 0x172a88ea in /usr/bin/clickhouse
5. ? @ 0x172a9e23 in /usr/bin/clickhouse
6. DB::commitMetadata(std::__1::function<void ()> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0x172de629 in /usr/bin/clickhouse
7. DB::MaterializeMetadata::transaction(DB::MySQLReplication::Position const&, std::__1::function<void ()> const&) @ 0x172e061a in /usr/bin/clickhouse
8. DB::MaterializeMySQLSyncThread::prepareSynchronized(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0x172a2fa1 in /usr/bin/clickhouse
9. DB::MaterializeMySQLSyncThread::synchronization(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0x172a6c5b in /usr/bin/clickhouse
10. ? @ 0x172a7176 in /usr/bin/clickhouse
11. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xff3e4a7 in /usr/bin/clickhouse
12. ? @ 0xff3cae3 in /usr/bin/clickhouse
13. start_thread @ 0x7dd5 in /usr/lib64/libpthread-2.17.so
14. __clone @ 0xfdead in /usr/lib64/libc-2.17.so

2020.08.28 00:01:42.719942 [ 7739 ] {cf4473a4-8b5a-4c3f-8aab-5f642e9bbf03} <Debug> MemoryTracker: Peak memory usage (for query): 0.00 B.

**Comment by BohuTANG** [comment]
@Enmk is working on this https://github.com/ClickHouse/ClickHouse/pull/11512

**Comment by liangli1213** [comment]
@BohuTANG ，I have one question, suppose we need to replicate 2 DB from same MySQL, will the bin log read twice by ClickHouse, for each of the DB?

**Comment by BohuTANG** [comment]
> @BohuTANG ，I have one question, suppose we need to replicate 2 DB from same MySQL, will the bin log read twice by ClickHouse, for each of the DB?

Yes, each database replication channel is a full streaming binlog network.
For multiple databases from the same MySQL maybe we can do it in MaterializeMySQL settings.

**Comment by lostsnow** [comment]
`create database planet_prod engine=MaterializeMySQL(...)`

> ClickHouse exception, code: 390, host: xxx, port: 8123; Code: 390, e.displayText() = DB::Exception: There is no SHOW CREATE TABLE query for DatabaseMaterializeMySQL (version 20.9.3.45 (official build))

https://paste.ubuntu.com/p/t8W95QkDZp/

`SELECT * FROM xxx LIMIT 50`

> Code: 50, e.displayText() = DB::Exception: Unsupported type Decimal(20, 4) (version 20.9.3.45 (official build)) 

https://paste.ubuntu.com/p/bnpRxCMyTg/

@BohuTANG

**Comment by tavplubix** [comment]
> `create database planet_prod engine=MaterializeMySQL(...)`
> 
> > ClickHouse exception, code: 390, host: xxx, port: 8123; Code: 390, e.displayText() = DB::Exception: There is no SHOW CREATE TABLE query for DatabaseMaterializeMySQL (version 20.9.3.45 (official build))

Fixed in #14849, not merged yet

> `SELECT * FROM xxx LIMIT 50`
> 
> > Code: 50, e.displayText() = DB::Exception: Unsupported type Decimal(20, 4) (version 20.9.3.45 (official build))

Decimals are available for MaterializeMySQL since 20.10
